contracts/src/TeleporterTokenSource.sol

```diff
@@ -206,56 +386,96 @@ abstract contract TeleporterTokenSource is
         bytes memory message
     ) internal override {
         BridgeMessage memory bridgeMessage = abi.decode(message, (BridgeMessage));
+        DestinationBridgeSettings memory destinationSettings =
+            registeredDestinations[sourceBlockchainID][originSenderAddress];
 
-        // Check that bridge instance returning has sufficient amount in balance
-        uint256 senderBalance = bridgedBalances[sourceBlockchainID][originSenderAddress];
-        require(
-            senderBalance >= bridgeMessage.amount,
-            "TeleporterTokenSource: insufficient bridge balance"
-        );
-
-        // Decrement the bridge balance by the unwrap amount
-        bridgedBalances[sourceBlockchainID][originSenderAddress] =
-            senderBalance - bridgeMessage.amount;
-
+        // TODO: do we need to check whether we received from a bridge that is registered?
+        // For receiving single transfers, this check is mostly handled by bridgeBalances,
+        // except if sent back 0 amount, which will still call _withdraw.
+        // For multi-hop transfers, bridgeBalances still checks and allows 0 amount to
+        // send a multi-hop. Then in prepareMultiHopRouting, it has the handling of 0 amount.
```

Comment:
Hmm, good call out. IIUC the questions here are:
- Should we allow for receiving/forwarding messages for 0 amount?
- If yes, should we allow for receiving messages for 0 amount from unregistered destinations?

If the first answer is no, then I'd be in favor of only adding the check that the amount is not 0, rather than having a redundant check on the destination registration and also having to check the amount.

If the first answer is yes, but the second answer is no, then we need to add that check as you mentioned.

If both answers are yes, then it sounds like we are fine as is.

Reply:
I don't think the answers themselves are too consequential either way. Currently, I don't think any of the destination contract implementations allow for sending an amount of 0, so to keep consistency I think it could make sense to disallow receiving an amount of 0. 

The only case I could potentially think of where that may be valid is when using `sendAndCall`, but I think it's reasonable to require the send amount to be non-zero there as well (if you're not sending any tokens, then you should just make the call through Teleporter directly).

So my slight preference would be to disallow receiving 0 amounts in all cases.

Reply:
Updated so all transfer messages have a `_processReceiverTransfer` call, which checks that the sending destination bridge is registered and collateralized, and also that the removed scaling source token amount is greater than zero to avoid withdrawing zero amount, though the destination bridges implement the last check by default too.